tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/Logging/NLog/LogsInjection/DiagnosticContextHelper.cs

```diff
@@ -127,12 +115,52 @@ public static void RemoveMdcState(MappedDiagnosticsContextSetterProxy mdc, bool
             }
         }
 
+        private static IReadOnlyList<KeyValuePair<string, object>> CreateEntriesList(Tracer tracer, out bool hasTraceIds)
+        {
+            hasTraceIds = false;
+            var spanContext = tracer.DistributedSpanContext;
+
+            if (spanContext is not null)
+            {
+                // For mismatch version support we need to keep requesting old keys.
+                var hasTraceId = spanContext.TryGetValue(SpanContext.Keys.TraceId, out string traceId) ||
+                                 spanContext.TryGetValue(HttpHeaderNames.TraceId, out traceId);
+                var hasSpanId = spanContext.TryGetValue(SpanContext.Keys.ParentId, out string spanId) ||
+                                spanContext.TryGetValue(HttpHeaderNames.ParentId, out spanId);
+                if (hasTraceId && hasSpanId)
+                {
+                    hasTraceIds = true;
+                    return new[]
+                    {
+                        new KeyValuePair<string, object>(CorrelationIdentifier.ServiceKey, tracer.DefaultServiceName ?? string.Empty),
+                        new KeyValuePair<string, object>(CorrelationIdentifier.VersionKey, tracer.Settings.ServiceVersion ?? string.Empty),
+                        new KeyValuePair<string, object>(CorrelationIdentifier.EnvKey, tracer.Settings.Environment ?? string.Empty),
+                        new KeyValuePair<string, object>(CorrelationIdentifier.TraceIdKey, traceId),
+                        new KeyValuePair<string, object>(CorrelationIdentifier.SpanIdKey, spanId)
+                    };
+                }
+            }
+
+            return new[]
+            {
+                new KeyValuePair<string, object>(CorrelationIdentifier.ServiceKey, tracer.DefaultServiceName ?? string.Empty),
+                new KeyValuePair<string, object>(CorrelationIdentifier.VersionKey, tracer.Settings.ServiceVersion ?? string.Empty),
+                new KeyValuePair<string, object>(CorrelationIdentifier.EnvKey, tracer.Settings.Environment ?? string.Empty)
+            };
+        }
+
         internal static class Cache<TMarker>
         {
             static Cache()
             {
                 var nlogAssembly = typeof(TMarker).Assembly;
 
+                if (GetScopeContextProxy(nlogAssembly) is { } scProxy)
+                {
+                    ScopeContext = scProxy;
+                    return;
+                }
+
```

Comment:
Does `NLog.ScopeContext` exist in all our supported versions of `NLog`? If not, then this would break log injection for those earlier versions? 

Similarly, if we don't need the `MdlcProxy` and `MdcProxy` in NLog 5.0 (because we've removed the integrations), then we could stop creating those integrations to improve startup perf?

Reply:
> Does NLog.ScopeContext exist in all our supported versions of NLog? If not, then this would break log injection for those earlier versions?

Not supported before v5, ScopeContext was released in 5.0.0. Log injection doesn't fail because GetXProxy methods try to create types in the reverse order they were introduced. In NLog MDC was first used, then MDLC, which was finally superseded by ScopeContext. `Cache` here looks in the opposite order (SC->MDLC->MDC), and only one of those ends up being created.

It is then a performance hit on startup for everything that is not ScopeContext. Options I can think about
1. Somehow pass the NLog version to the Cache to go straight into one specific class
2. Create a different Cache to use in v5 only

Not sure if it's worth it for something that should be used less each day, as more customers adopt NLog 5 or higher. [Usage of v4 seem to still be pretty high](https://www.nuget.org/stats/packages/NLog?groupby=Version) though.

WDYT?